pybmix/core/pybmixcpp/python_embedding/python_hierarchy.cc

```diff
@@ -0,0 +1,444 @@
+                    state.generic_state, sum_stats, py_gen, cluster_data_values,
+                    hypers->generic_hypers);
+            synchronize_py_to_cpp_state(bayesmix::Rng::Instance().get(), py_gen);
+            py::list state_list = result[0];
+            py::list sum_stats_list = result[1];
+            state.generic_state = list_to_vector(state_list);
+            sum_stats = list_to_vector(sum_stats_list);
+        }
+    }
+}
+
+void PythonHierarchy::sample_full_cond(
+        const Eigen::MatrixXd &data,
+        const Eigen::MatrixXd &covariates /*= Eigen::MatrixXd(0, 0)*/) {
+    clear_data();
+    clear_summary_statistics();
+    if (covariates.cols() == 0) {
+        // Pass null value as covariate
+        for (int i = 0; i < data.rows(); i++) {
+            (this)->add_datum(i, data.row(i), false, Eigen::RowVectorXd(0));
+        }
+    } else if (covariates.rows() == 1) {
+        // Use unique covariate
+        for (int i = 0; i < data.rows(); i++) {
+            (this)->add_datum(i, data.row(i), false, covariates.row(0));
+        }
+    } else {
+        // Use different covariates
+        for (int i = 0; i < data.rows(); i++) {
+            (this)->add_datum(i, data.row(i), false, covariates.row(i));
+        }
+    }
+    (this)->sample_full_cond(true);
+}
+
+void PythonHierarchy::sample_prior() { state = (this)->draw(*hypers); };
+
+void PythonHierarchy::save_posterior_hypers() {
+    if (this->is_conjugate()) {
+        posterior_hypers = (this)->compute_posterior_hypers();
+    } else {
+        throw std::runtime_error("save_posterior_hypers() not implemented");
+    }
+};
+
+//! PYTHON
+void PythonHierarchy::update_hypers(
+        const std::vector <bayesmix::AlgorithmState::ClusterState> &states) {
+    // std::cout << "Starting update_hypers!" << std::endl;
+    unsigned int size_states = states.size();
+    std::vector <std::vector<double>> pass_states{};
+    pass_states.reserve(size_states);
+    unsigned int size;
+    for (auto &st: states) {
+        size = st.general_state().size();
+        std::vector<double> aux_v{};
+        aux_v.reserve(size);
+        for (int i = 0; i < size; ++i) {
+            aux_v.push_back((st.general_state().data())[i]);
+        }
+        pass_states.push_back(aux_v);
+    }
+    synchronize_cpp_to_py_state(bayesmix::Rng::Instance().get(), py_gen);
+    py::list new_hypers =
+            update_hypers_evaluator(pass_states, hypers->generic_hypers, py_gen);
+    synchronize_py_to_cpp_state(bayesmix::Rng::Instance().get(), py_gen);
+    hypers->generic_hypers = list_to_vector(new_hypers);
+}
+
+void PythonHierarchy::add_datum(
+        const int id, const Eigen::RowVectorXd &datum,
+        const bool update_params /*= false*/,
+        const Eigen::RowVectorXd &covariate /*= Eigen::RowVectorXd(0)*/) {
+    assert(cluster_data_idx.find(id) == cluster_data_idx.end());
+    card += 1;
+    log_card = std::log(card);
```

Comment:
This should probably be `set_card(card + 1);`

Reply:
`set_card()` also takes care of the logarithm, so you also have to remove the `log_card` assignment